tensorflow/core/kernels/training_ops.cc

```diff
@@ -1003,32 +1037,41 @@ REGISTER_KERNELS(GPU, complex128);
 #undef REGISTER_CPU_KERNELS
 #undef REGISTER_KERNELS
 
-// Note, this op works on cpu only.
-template <typename T, typename Tindex>
+template <typename T, typename Device, typename Tindex>
 class SparseApplyAdadeltaOp : public OpKernel {
  public:
   explicit SparseApplyAdadeltaOp(OpKernelConstruction* ctx) : OpKernel(ctx) {
     OP_REQUIRES_OK(ctx, ctx->GetAttr("use_locking", &use_exclusive_lock_));
   }
 
   void Compute(OpKernelContext* ctx) override {
+    Var* var;
     const bool sparse = true;
-    auto locks = MaybeLockVariableInputMutexesInOrder<CPUDevice, T>(
-        ctx, use_exclusive_lock_, sparse, {0, 1, 2});
-    DoCompute(ctx);
+    mutex* mu = GetTrainingVariableMutex<Device, T>(ctx, 0, sparse, &var);
+    core::ScopedUnref scoped_unref(var);
+    // mu_accum is actually the same mutex as mu_var since currently we use a
+    // global mutex.
+    //
+    // mutex* mu_accum = ctx->input_ref_mutex(1);
```

Comment:
How about the mutex for `accum_update`?

Also, I don't really understand the comment; if `accum` and `var` are different ref variables then won't they have different mutexes?

Reply:
I have changed the logic of this part.